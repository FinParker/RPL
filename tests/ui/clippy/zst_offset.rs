//@revisions: inline normal
//@compile-flags: -Z mir-opt-level=0
//@[inline]compile-flags: -Z inline-mir=true
//@[normal]compile-flags: -Z inline-mir=false

// #[rpl::dump_mir(dump_cfg, dump_ddg)]
fn main() {
    unsafe {
        let m = &mut () as *mut ();
        m.offset(0);
        //~^ zst_offset

        m.wrapping_add(0);
        //~^ zst_offset

        m.sub(0);
        //~^ zst_offset

        m.wrapping_sub(0);
        //~^ zst_offset

        let c = &() as *const ();
        c.offset(0);
        //~^ zst_offset
        //~[inline]| unchecked_pointer_offset

        c.wrapping_add(0);
        //~^ zst_offset

        c.sub(0);
        //~^ zst_offset
        //~[inline]| unchecked_pointer_offset

        c.wrapping_sub(0);
        //~^ zst_offset

        let sized = &1 as *const i32;
        sized.offset(0);
        //~[inline]^ unchecked_pointer_offset
    }
}
