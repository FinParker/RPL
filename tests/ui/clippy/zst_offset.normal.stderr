error: offset calculation on zero-sized value
  --> tests/ui/clippy/zst_offset.rs:23:9
   |
LL |         c.offset(0);
   |         ^^^^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address
   = note: `#[deny(rpl::zst_offset)]` on by default

error: offset calculation on zero-sized value
  --> tests/ui/clippy/zst_offset.rs:10:9
   |
LL |         m.offset(0);
   |         ^^^^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address

error: offset calculation on zero-sized value
  --> tests/ui/clippy/zst_offset.rs:29:9
   |
LL |         c.sub(0);
   |         ^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address

error: offset calculation on zero-sized value
  --> tests/ui/clippy/zst_offset.rs:16:9
   |
LL |         m.sub(0);
   |         ^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address

error: offset calculation on zero-sized value
  --> tests/ui/clippy/zst_offset.rs:26:9
   |
LL |         c.wrapping_add(0);
   |         ^^^^^^^^^^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address

error: offset calculation on zero-sized value
  --> tests/ui/clippy/zst_offset.rs:13:9
   |
LL |         m.wrapping_add(0);
   |         ^^^^^^^^^^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address

error: offset calculation on zero-sized value
  --> tests/ui/clippy/zst_offset.rs:32:9
   |
LL |         c.wrapping_sub(0);
   |         ^^^^^^^^^^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address

error: offset calculation on zero-sized value
  --> tests/ui/clippy/zst_offset.rs:19:9
   |
LL |         m.wrapping_sub(0);
   |         ^^^^^^^^^^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address

error: aborting due to 8 previous errors

